src/uct/ib/ud/base/ud_ep.c

```diff
@@ -453,15 +459,40 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
     ucs_mpool_put(skb);
 }
 
+ucs_status_t uct_ud_ep_flush_do(uct_ud_iface_t *iface, uct_ud_ep_t *ep)
+{
+    uct_ud_send_skb_t *skb;
+
+    if (ucs_queue_is_empty(&ep->tx.window)) {
+        uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_ACK_REQ);
+        return UCS_OK;
+    }
+    
+    skb = ucs_queue_tail_elem_non_empty(&ep->tx.window, uct_ud_send_skb_t, queue);
+    if (skb->neth->packet_type & UCT_UD_PACKET_FLAG_ACK_REQ) {
+        /* last packet was already sent with ack request. 
+         * either by flush or 
+         * Do not send more, let reqular retransmission 
+         * mechanism do the work
+         */
+        return UCS_INPROGRESS;
+    }
+
+    skb->neth->packet_type |= UCT_UD_PACKET_FLAG_ACK_REQ;
```

Comment:
also, probably should add valgrind macros to post_send/send_completion to catch this